File: src/main/java/com/donrobo/fpbg/blueprint/Blueprint.java
package com.donrobo.fpbg.blueprint;

import com.donrobo.fpbg.blueprint.building.Building;
import com.donrobo.fpbg.data.Int2;
import net.sf.json.JSONArray;
import net.sf.json.JSONObject;
import org.apache.commons.io.IOUtils;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.zip.DeflaterOutputStream;

public class Blueprint {

    private final List<Building> buildings = new ArrayList<>();

    public JSONObject toJson() {
        JSONObject json = new JSONObject();

        JSONObject blueprintJson = new JSONObject();
        JSONArray entitiesArray = new JSONArray();

        int i = 0;
        for (Building building : buildings) {
            i++;
            JSONObject buildingJson = building.toJson();
            buildingJson.put("entity_number", i);
            entitiesArray.add(buildingJson);
        }

        blueprintJson.put("entities", entitiesArray);
        json.put("blueprint", blueprintJson);
        json.put("item", "blueprint");

        return json;
    }

    public static String toBlueprintString(String input) {
        ByteArrayOutputStream baos = null;
        DeflaterOutputStream dos = null;

        try {
            baos = new ByteArrayOutputStream();
            dos = new DeflaterOutputStream(baos);

            dos.write(input.getBytes(Charset.forName("UTF-8")));
            IOUtils.closeQuietly(dos);

            return "0" + Base64.getEncoder().encodeToString(baos.toByteArray());
        } catch (IOException ex) {
            throw new RuntimeException("Shouldn't happen here", ex);
        } finally {
            IOUtils.closeQuietly(baos);
            IOUtils.closeQuietly(dos);
        }
    }

    public String toBlueprintString() {
        return toBlueprintString(toJson().toString());
    }

    public void addBuilding(Building building) {
        if (isOccupied(building)) {
            throw new RuntimeException("Trying to place building on top of building!");
        }
        buildings.add(building);
    }

    public boolean isOccupied(Int2 pos) {
        return isOccupied(pos.getX(), pos.getY());
    }

    public boolean isOccupied(int x, int y) {
        return get(x, y) != null;
    }

    public boolean isOccupied(int x, int y, int width, int height) {
        for (int xI = x; xI < x + width; xI++) {
            for (int yI = y; yI < y + height; yI++) {
                if (isOccupied(xI, yI)) {
                    return true;
                }
            }
        }

        return false;
    }

    public boolean isOccupied(Building building) {
        return isOccupied(building.getX(), building.getY(), building.getWidth(), building.getHeight());
    }

    public Building get(int x, int y) {
        return buildings.stream().filter(building -> {
            boolean xFits = building.getX() <= x && building.getX() + building.getWidth() - 1 >= x;
            boolean yFits = building.getY() <= y && building.getY() + building.getHeight() - 1 >= y;

            return xFits && yFits;
        }).findAny().orElse(null);
    }

    public Building remove(int x, int y) {
        Building building = get(x, y);
        if (building != null) {
            buildings.remove(building);
        }
        return building;
    }

    public void addBlueprint(Blueprint blueprint, int xOffset, int yOffset) {
        blueprint.buildings.stream().map(b -> b.move(xOffset, yOffset)).forEach(this::addBuilding);
    }

    public int getMinimumX() {
        return buildings.stream().mapToInt(Building::getX).min().orElse(0);
    }

    public int getMaximumX() {
        return buildings.stream().mapToInt(building -> building.getX() + building.getWidth() - 1).max().orElse(0);
    }

    public int getMinimumY() {
        return buildings.stream().mapToInt(Building::getY).min().orElse(0);
    }

    public int getMaximumY() {
        return buildings.stream().mapToInt(building -> building.getY() + building.getHeight() - 1).max().orElse(0);
    }

    public int getWidth() {
        return getMaximumX() - getMinimumX();
    }

    public int getHeight() {
        return getMaximumY() - getMinimumY();
    }

    public boolean canPlace(Blueprint blueprint, int x, int y) {
        for (Building building : blueprint.buildings) {
            if (isOccupied(building.move(x, y))) {
                return false;
            }
        }
        return true;
    }

    public List<Building> getBuildings() {
        return buildings;
    }

    public Building get(Int2 pos) {
        return get(pos.getX(), pos.getY());
    }
}

File: src/main/kotlin/com/donrobo/fpbg/blueprint/Blueprint.kt
package com.donrobo.fpbg.blueprint

import com.donrobo.fpbg.blueprint.Direction.*
import com.donrobo.fpbg.blueprint.building.Building
import com.donrobo.fpbg.blueprint.building.YellowBelt
import com.donrobo.fpbg.data.Int2
import com.donrobo.fpbg.util.Map2D
import com.donrobo.fpbg.util.asMap
import net.sf.json.JSONArray
import net.sf.json.JSONObject
import org.apache.commons.io.IOUtils
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.nio.charset.Charset
import java.util.*
import java.util.zip.DeflaterOutputStream

class Blueprint {

    private val internalBuildings = ArrayList<Building>()

    fun toJson(): JSONObject {
        val json = JSONObject()

        val blueprintJson = JSONObject()
        val entitiesArray = JSONArray()

        var i = 0
        for (building in buildings) {
            i++
            val buildingJson = building.toJson()
            buildingJson.put("entity_number", i)
            entitiesArray.add(buildingJson)
        }

        blueprintJson.put("entities", entitiesArray)
        json.put("blueprint", blueprintJson)
        json.put("item", "blueprint")

        return json
    }

    fun toBlueprintString(): String {
        return toBlueprintString(toJson().toString())
    }

    fun addBuilding(building: Building) {
        if (isOccupied(building)) {
            throw RuntimeException("Trying to place building on top of building!")
        }
        internalBuildings.add(building)
    }

    fun isOccupied(pos: Int2) = isOccupied(pos.x, pos.y)

    fun isOccupied(x: Int, y: Int) = get(x, y) != null

    fun isOccupied(x: Int, y: Int, width: Int, height: Int): Boolean {
        for (xI in x..(x + width - 1)) {
            (y..y + height - 1).filter { isOccupied(xI, it) }.forEach { return true }
        }

        return false
    }

    fun isOccupied(building: Building) = isOccupied(building.x, building.y, building.width, building.height)

    operator fun get(x: Int, y: Int): Building? = buildings.filter { building ->
        val xFits = building.x <= x && building.x + building.width - 1 >= x
        val yFits = building.y <= y && building.y + building.height - 1 >= y

        xFits && yFits
    }.firstOrNull()

    fun remove(x: Int, y: Int): Building? {
        val building = get(x, y)
        if (building != null) {
            internalBuildings.remove(building)
        }
        return building
    }

    fun addBlueprint(blueprint: Blueprint, xOffset: Int = 0, yOffset: Int = 0) {
        blueprint.buildings.map { b -> b.move(xOffset, yOffset) }.forEach { this.addBuilding(it) }
    }

    fun move(xOffset: Int, yOffset: Int): Blueprint {
        val blueprint = Blueprint()

        blueprint.addBlueprint(this, xOffset, yOffset)

        return blueprint
    }

    val minimumX: Int
        get() = buildings.map { it.x }.min() ?: 0

    val maximumX: Int
        get() = buildings.map { it.x + it.width - 1 }.max() ?: 0

    val minimumY: Int
        get() = buildings.map { it.y }.min() ?: 0

    val maximumY: Int
        get() = buildings.map { it.y + it.height - 1 }.max() ?: 0

    val width: Int
        get() = maximumX - minimumX + 1//1-1 is still one width

    val height: Int
        get() = maximumY - minimumY + 1//1-1 is still one height

    fun canPlace(blueprint: Blueprint, x: Int, y: Int): Boolean {
        return blueprint.buildings.none { isOccupied(it.move(x, y)) }
    }

    val buildings: List<Building> get() = internalBuildings

    operator fun get(pos: Int2): Building? {
        return get(pos.x, pos.y)
    }

    fun visualizer(): Map2D {
        val mapVisualizer = Map2D()

        for (building in buildings) {
            for (y in building.y..(building.y + building.height - 1)) {
                for (x in building.x..(building.x + building.width - 1)) {
                    mapVisualizer[Int2(x, y)] = building.visualizationCharacter
                }
            }
        }

        return mapVisualizer
    }

    companion object {

        fun toBlueprintString(input: String): String {
            var baos: ByteArrayOutputStream? = null
            var dos: DeflaterOutputStream? = null

            try {
                baos = ByteArrayOutputStream()
                dos = DeflaterOutputStream(baos)

                dos.write(input.toByteArray(Charset.forName("UTF-8")))
                IOUtils.closeQuietly(dos)

                return "0" + Base64.getEncoder().encodeToString(baos.toByteArray())
            } catch (ex: IOException) {
                throw RuntimeException("Shouldn't happen here", ex)
            } finally {
                IOUtils.closeQuietly(baos)
                IOUtils.closeQuietly(dos)
            }
        }
    }

    fun visualize(showCoordinates: Boolean = false): String {
        return visualizer().visualize(showCoordinates)
    }

    fun rotateCW(around: Int2 = Int2(0, 0), count: Int): Blueprint {
        val rotated = Blueprint()

        for (building in buildings) {
            rotated.addBuilding(building.rotateCW(around, count))
        }

        return rotated
    }
}

fun String.toBeltBlueprint(): Blueprint {
    return asMap().toBeltBlueprint()
}

fun Map2D.toBeltBlueprint(): Blueprint {
    val blueprint = Blueprint()

    asMap().forEach { (pos, char) ->
        if (char != ' ')
            blueprint.addBuilding(when (char) {
                '^' -> YellowBelt(pos.x, pos.y, UP)
                '>' -> YellowBelt(pos.x, pos.y, RIGHT)
                'v' -> YellowBelt(pos.x, pos.y, DOWN)
                '<' -> YellowBelt(pos.x, pos.y, LEFT)
                else -> throw IllegalArgumentException("$char is not supported")
            })

    }

    assert(width == blueprint.width)
    assert(height == blueprint.height)

    return blueprint
}
